modules/layers/src/mode-handlers/snap-handler.js

```diff
@@ -0,0 +1,304 @@
+// @flow
+
+import turfArea from '@turf/area';
+import turfBearing from '@turf/bearing';
+import turfCenter from '@turf/center';
+import turfDistance from '@turf/distance';
+import turfEnvelope from '@turf/envelope';
+import { point as turfPoint } from '@turf/helpers';
+import turfTransformTranslate from '@turf/transform-translate';
+
+import type { FeatureCollection, Position } from '../geojson-types.js';
+import { SortedList, SortType } from '../utils';
+import type { EditHandle } from './mode-handler';
+import { ModeHandler, getEditHandlesForGeometry } from './mode-handler';
+
+export class SnapHandler extends ModeHandler {
```

Comment:
I'm not sure we need this class still. If you look at the pseudocode I posted, I think it may be pretty trivial to get the gray coordinate close to the cursor and use the existing `getTranslateAction` to translate from where the current purple coordinate is to the where the gray coordinate is.